sys/random/minstd.c

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ * Code taken from C FAQ (http://c-faq.com/lib/rand.html).
+ */
+
+ /**
+ * @ingroup sys_random
+ * @{
+ * @file
+ *
+ * @brief Simple Park & Miller "minimal standard" PRNG
+ *
+ * This file contains a simple Park-Miller pseudo random number generator.
+ *
+ * While not very random when considering crypto requirements, this is probably
+ * random enough anywhere where pseudo-randomness is sufficient, e.g., when
+ * provided with a sensible seed source, for MAC algorithms.
+ *
+ * The implementation is taken from the C FAQ, but modified to use magic number
+ * division and adapted to RIOT's coding conventions..
+ *
+ * @author Kaspar Schleiser <kaspar@schleiser.de>
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "div.h"
+
+#define a 48271
+#define m 2147483647
+#define q (m / a) /* 44488 */
+#define r (m % a) /* 3399 */
+
+static uint32_t _seed = 1;
+
+static uint32_t PMrand(void)
+{
+    uint32_t hi = div_u32_by_44488(_seed);
+    uint32_t lo = div_u32_mod_44488(_seed);
+    uint32_t test = (a * lo) - (r * hi);
+
+    if(test > 0) {
+        _seed = test;
+    }
+    else {
+        _seed = test + m;
+    }
+
+    return _seed;
+}
+
+uint32_t genrand_uint32(void)
+{
+    /* PMrand() only returns values from 1 to 2147483647, so in order to
+     * get full 32bit values, just add two. */
+    return PMrand() + PMrand();
```

Comment:
Also missing is 4294967295. It is a problem if the user is expecting the full range of the uint32_t type. Using two more calls to PMrand would make the function able to return all values, but 0, 1, 4294967294, and 4294967295 would be less likely than the other numbers (if I got the modulo math reasoning right)